integration_test/index_test.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2019 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integrationtest
+
+import (
+	"bytes"
+	"os"
+	"testing"
+
+	"sigs.k8s.io/krew/pkg/constants"
+)
+
+func TestKrewIndexAdd(t *testing.T) {
+	skipShort(t)
+
+	test, cleanup := NewTest(t)
+	defer cleanup()
+
+	test.WithEnv(constants.EnableMultiIndexSwitch, 1).WithIndex()
+	if err := test.Krew("index", "add").Run(); err == nil {
+		t.Fatal("expected index add with no args to fail")
+	}
+	if err := test.Krew("index", "add", "foo", "https://invalid").Run(); err == nil {
+		t.Fatal("expected index add with invalid URL to fail")
+	}
+	if err := test.Krew("index", "add", "../../usr/bin", constants.IndexURI); err == nil {
+		t.Fatal("expected index add with path characters to fail")
+	}
+	if err := test.Krew("index", "add", "foo", constants.IndexURI).Run(); err != nil {
+		t.Fatalf("error adding new index: %v", err)
+	}
+}
+
+func TestKrewIndexList(t *testing.T) {
+	skipShort(t)
+
+	test, cleanup := NewTest(t)
+	defer cleanup()
+
+	test.WithEnv(constants.EnableMultiIndexSwitch, 1).WithIndex()
+	out := test.Krew("index", "list").RunOrFailOutput()
+	if !bytes.Contains(out, []byte(constants.DefaultIndexName)) {
+		t.Fatalf("expected index 'default' in output:\n%s", string(out))
+	}
+
+	test.Krew("index", "add", "foo", constants.IndexURI).RunOrFail()
+	out = test.Krew("index", "list").RunOrFailOutput()
+	if !bytes.Contains(out, []byte("foo")) {
+		t.Fatalf("expected index 'foo' in output:\n%s", string(out))
+	}
+}
+
+func TestKrewIndexList_NoIndexes(t *testing.T) {
+	skipShort(t)
+
+	test, cleanup := NewTest(t)
+	defer cleanup()
+
+	test.WithEnv(constants.EnableMultiIndexSwitch, 1).WithIndex()
+	index := test.TempDir().Path("index/" + constants.DefaultIndexName)
```

Comment:
you can just remove `index` completely. that should still return empty list, right?

Reply:
List uses `paths.IndexBase()` right now so it returns an error when it tries to `ReadDir()` it to see what indexes are there.

Reply:
Hmm I'm not sure it should fail in case of os.IsNotExist...

What if user has never run krew (so ~/.krew doesn't even exist). IMO `krew index list` should not fail. @corneliusweig right?

Reply:
Yeah, this is a subtle bug... The problem is that the index is only initialized for _some_ commands (install, update, upgrade), see `ensureIndexUpdated`. So if the index is removed, our index migration check thinks it is already migrated and does nothing. This is not some construed case, because it is prefectly reasonable to run `krew index list` right after install. So we need to fix this.

And when it's done, it should be enough to remove `index` as suggested.

Reply:
So if a user has 0.3.4 and has never run krew they won't have access to the index command until they update krew. If they update through krew they'll have old index directory and next command will prompt to migrate (or automatically migrate if/when we change to that).

Fresh install of newer version means the user is already technically "migrated" even without the index directory being initialized right? The default krew-index should be initialized in the new structure, or a different "default" can be added through `krew index add`.

Do those sound right to you guys or is there a scenario I'm missing? As to `index list` not failing, I think its just a matter of not treating `os.IsNotExist` as a failure like Ahmet mentioned.

Reply:
> Fresh install of newer version means the user is already technically "migrated" even without the index directory being initialized right? The default krew-index should be initialized in the new structure, or a different "default" can be added through `krew index add`.

I am hoping we solve this in a different PR.

For this PR: when Krew runs, it will always go ahead and create some dirs in ~/.krew. https://github.com/kubernetes-sigs/krew/blob/5c18854cb38e5a258bb73015391eed49a6c99932/cmd/krew/cmd/root.go#L110-L114 Right now it doesn't have `index/` dir, so let's go ahead and add it there.

That way, any time krew runs (even for --help or a failure) the `index/` will exist. That way, we can detect no index exists (ls index/ == 0 items) and initialize index.

Reply:
I guess, creating `index/` also fixes the bug entirely.